dhis-2/dhis-e2e-test/src/main/java/org/hisp/dhis/actions/SystemActions.java

```diff
@@ -46,30 +51,52 @@ public SystemActions()
         super( "/system" );
     }
 
+    public ApiResponse getTask( String taskType, String taskId )
+    {
+        return get( "/tasks/" + taskType + "/" + taskId );
+    }
+
+    /**
+     * Waits until the task is completed and returns a response
+     *
+     * @param taskType
+     * @param taskId
+     * @return
+     */
     public ApiResponse waitUntilTaskCompleted( String taskType, String taskId )
     {
-        logger.info( "Waiting until task " + taskType + " with id " + taskId + "is completed" );
-        ApiResponse response = null;
-        boolean completed = false;
-        while ( !completed )
-        {
-            response = get( "/tasks/" + taskType + "/" + taskId );
-            response.validate().statusCode( 200 );
-            completed = response.extractList( "completed" ).contains( true );
-        }
+        Callable<Boolean> taskIsCompleted = () -> getTask( taskType, taskId )
+            .validateStatus( 200 )
+            .extractList( "completed" ).contains( true );
+
+        with()
+            .pollInterval( 1, TimeUnit.SECONDS )
+            .atMost( 20, TimeUnit.SECONDS )
+            .await().until( () -> taskIsCompleted.call() );
 
-        logger.info( "Task completed. Message: " + response.extract( "message" ) );
-        return response;
+        return getTask( taskType, taskId );
     }
 
     public List<ImportSummary> getTaskSummaries( String taskType, String taskId )
     {
-        return getTaskSummariesResponse( taskType, taskId ).validateStatus( 200 ).getImportSummaries();
+        return waitForTaskSummaries( taskType, taskId ).validateStatus( 200 ).getImportSummaries();
     }
 
-    public ApiResponse getTaskSummariesResponse( String taskType, String taskId )
+    /**
+     * Waits until task summaries are generated and returns the response
+     *
+     * @param taskType
+     * @param taskId
+     * @return
+     */
+    public ApiResponse waitForTaskSummaries( String taskType, String taskId )
     {
-        return get( "/taskSummaries/" + taskType + "/" + taskId );
+        String url = String.format( "/taskSummaries/%s/%s", taskType, taskId );
+
+        await().ignoreExceptions().until( () -> !get( url ).validateStatus( 200 )
```

Comment:
I think this now has a potential to never finish in case you never get a 200. We should add some timeout.

Reply:
@jbee Awatility's default timeout is 10 seconds. [[ref](https://github.com/awaitility/awaitility/wiki/Usage#defaults)]. I can set it explicitly if that's clearer.

Reply:
Ah, ok - as long as there is a timeout I am fine with this. No need to set it explicitly.